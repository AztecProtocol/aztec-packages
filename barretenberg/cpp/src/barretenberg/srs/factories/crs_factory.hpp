#pragma once
#include "barretenberg/common/mem.hpp"
#include "barretenberg/ecc/curves/bn254/bn254.hpp"
#include "barretenberg/ecc/curves/bn254/g1.hpp"
#include "barretenberg/ecc/curves/bn254/g2.hpp"
#include "barretenberg/ecc/curves/grumpkin/grumpkin.hpp"
#include <cstddef>

namespace bb::pairing {
struct miller_lines;
} // namespace bb::pairing

namespace bb::srs::factories {

/**
 * A prover crs representation.
 */
template <typename Curve> class ProverCrs {
  public:
    virtual ~ProverCrs() = default;
    ;

    /**
     *  @brief Returns the monomial points in a form to be consumed by scalar_multiplication pippenger algorithm.
     */
    virtual typename Curve::AffineElement* get_monomial_points() = 0;
    virtual size_t get_monomial_size() const = 0;
};

template <typename Curve> class VerifierCrs {
  public:
    virtual ~VerifierCrs() = default;
};
template <> class VerifierCrs<curve::BN254> {
    using Curve = curve::BN254;

  public:
    virtual Curve::G2AffineElement get_g2x() const = 0;
    /**
     * @brief As the G_2 element of the CRS is fixed, we can precompute the operations performed on it during the
     * pairing algorithm to optimize pairing computations.
     */
    virtual bb::pairing::miller_lines const* get_precomputed_g2_lines() const = 0;
    /**
     *  @brief Returns the first G_1 element from the CRS, used by the Shplonk verifier to compute the final
     * commtiment.
     */
    virtual Curve::AffineElement get_first_g1() const = 0;
};

template <> class VerifierCrs<curve::Grumpkin> {
    using Curve = curve::Grumpkin;

  public:
    /**
     * @brief Returns the G_1 elements in the CRS after the pippenger point table has been applied on them
     *
     */
    virtual Curve::AffineElement* get_monomial_points() const = 0;
    virtual size_t get_monomial_size() const = 0;
    /**
     * @brief Returns the first G_1 element from the CRS, used by the Shplonk verifier to compute the final
     * commtiment.
     */
    virtual Curve::AffineElement get_first_g1() const = 0;
};

/**
 * A factory class to return the prover crs and verifier crs on request.
 * You can construct an empty placeholder factory, because composers need to be given a factory at construction time.
 */
template <typename Curve> class CrsFactory {
  public:
    CrsFactory() = default;
    CrsFactory(CrsFactory&& other) = default;
    virtual ~CrsFactory() = default;
    virtual std::shared_ptr<bb::srs::factories::ProverCrs<Curve>> get_prover_crs(size_t) { return nullptr; }
    virtual std::shared_ptr<bb::srs::factories::VerifierCrs<Curve>> get_verifier_crs([[maybe_unused]] size_t degree = 0)
    {
        return nullptr;
    }
};

} // namespace bb::srs::factories