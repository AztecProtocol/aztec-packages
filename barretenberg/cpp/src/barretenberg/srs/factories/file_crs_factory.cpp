#include "file_crs_factory.hpp"
#include "../io.hpp"
#include "barretenberg/ecc/curves/bn254/bn254.hpp"
#include "barretenberg/ecc/curves/bn254/g1.hpp"
#include "barretenberg/ecc/curves/bn254/pairing.hpp"
#include "barretenberg/ecc/curves/grumpkin/grumpkin.hpp"
#include "barretenberg/ecc/scalar_multiplication/point_table.hpp"
#include "barretenberg/ecc/scalar_multiplication/scalar_multiplication.hpp"

namespace bb::srs::factories {

FileVerifierCrs<curve::BN254>::FileVerifierCrs(std::string const& path, const size_t)
    : precomputed_g2_lines((bb::pairing::miller_lines*)(aligned_alloc(64, sizeof(bb::pairing::miller_lines) * 2)))
{
    using Curve = curve::BN254;
    auto point_buf = scalar_multiplication::point_table_alloc<Curve::AffineElement>(1);
    srs::IO<Curve>::read_transcript_g1(point_buf.get(), 1, path);
    srs::IO<curve::BN254>::read_transcript_g2(g2_x, path);
    bb::pairing::precompute_miller_lines(bb::g2::one, precomputed_g2_lines[0]);
    bb::pairing::precompute_miller_lines(g2_x, precomputed_g2_lines[1]);
    first_g1 = point_buf[0];
}

FileVerifierCrs<curve::BN254>::~FileVerifierCrs()
{
    aligned_free(precomputed_g2_lines);
}

FileVerifierCrs<curve::Grumpkin>::FileVerifierCrs(std::string const& path, const size_t num_points)
    : num_points(num_points)
{
    using Curve = curve::Grumpkin;
    monomials_ = scalar_multiplication::point_table_alloc<Curve::AffineElement>(num_points);
    srs::IO<Curve>::read_transcript_g1(monomials_.get(), num_points, path);
    scalar_multiplication::generate_pippenger_point_table<Curve>(monomials_.get(), monomials_.get(), num_points);
    first_g1 = monomials_[0];
};

curve::Grumpkin::AffineElement* FileVerifierCrs<curve::Grumpkin>::get_monomial_points() const
{
    return monomials_.get();
}

size_t FileVerifierCrs<curve::Grumpkin>::get_monomial_size() const
{
    return num_points;
}

template <typename Curve>
FileCrsFactory<Curve>::FileCrsFactory(std::string path, size_t initial_degree)
    : path_(std::move(path))
    , degree_(initial_degree)
{}

template <typename Curve>
std::shared_ptr<bb::srs::factories::ProverCrs<Curve>> FileCrsFactory<Curve>::get_prover_crs(size_t degree)
{
    if (degree != degree_ || !prover_crs_) {
        prover_crs_ = std::make_shared<FileProverCrs<Curve>>(degree, path_);
        degree_ = degree;
    }
    return prover_crs_;
}

template <typename Curve>
std::shared_ptr<bb::srs::factories::VerifierCrs<Curve>> FileCrsFactory<Curve>::get_verifier_crs(size_t degree)
{
    if (degree != degree_ || !verifier_crs_) {
        verifier_crs_ = std::make_shared<FileVerifierCrs<Curve>>(path_, degree);
        degree_ = degree;
    }
    return verifier_crs_;
}

template class FileProverCrs<curve::BN254>;
template class FileProverCrs<curve::Grumpkin>;
template class FileCrsFactory<curve::BN254>;
template class FileCrsFactory<curve::Grumpkin>;

} // namespace bb::srs::factories
