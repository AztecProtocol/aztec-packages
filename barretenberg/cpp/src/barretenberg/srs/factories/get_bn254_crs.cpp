#include "get_bn254_crs.hpp"
#include "barretenberg/api/exec_pipe.hpp"
#include "barretenberg/api/file_io.hpp"
#include "barretenberg/ecc/curves/bn254/g1.hpp"

namespace {
std::vector<uint8_t> download_bn254_g1_data(size_t num_points)
{
    size_t g1_end = num_points * sizeof(bb::g1::affine_element) - 1;

    std::string url = "https://crs.aztec.network/g1.dat";

    // IMPORTANT: this currently uses a shell, DO NOT let user-controlled strings here.
    std::string command = "curl -H \"Range: bytes=0-" + std::to_string(g1_end) + "\" '" + url + "'";

    auto data = bb::exec_pipe(command);
    // Header + num_points * sizeof point.
    if (data.size() < g1_end) {
        throw_or_abort("Failed to download g1 data.");
    }

    return data;
}

std::vector<uint8_t> download_bn254_g2_data()
{
    std::string url = "https://crs.aztec.network/g2.dat";
    // IMPORTANT: this currently uses a shell, DO NOT let user-controlled strings here.
    std::string command = "curl '" + url + "'";
    return bb::exec_pipe(command);
}
} // namespace

namespace bb {
std::vector<g1::affine_element> get_bn254_g1_data(const std::filesystem::path& path,
                                                  size_t num_points,
                                                  bool allow_download)
{
    // TODO(AD): per Charlie this should just download and replace the flat file portion atomically so we have no race
    // condition
    std::filesystem::create_directories(path);

    auto g1_path = path / "bn254_g1.dat";
    size_t g1_downloaded_points = get_file_size(g1_path) / sizeof(g1::affine_element);

    if (g1_downloaded_points >= num_points) {
        vinfo("using cached bn254 crs with num points ", std::to_string(g1_downloaded_points), " at ", g1_path);
        auto data = read_file(g1_path, num_points * sizeof(g1::affine_element));
        auto points = std::vector<g1::affine_element>(num_points);
        for (size_t i = 0; i < num_points; ++i) {
            points[i] = from_buffer<g1::affine_element>(data, i * sizeof(g1::affine_element));
        }
        return points;
    }

    if (!allow_download && g1_downloaded_points == 0) {
        throw_or_abort("bn254 g1 data not found and download not allowed in this context");
    } else if (!allow_download) {
        throw_or_abort(format("bn254 g1 data had ",
                              g1_downloaded_points,
                              " points and ",
                              num_points,
                              " were requested but download not allowed in this context"));
    }
    vinfo("downloading bn254 crs...");
    auto data = download_bn254_g1_data(num_points);
    write_file(g1_path, data);

    auto points = std::vector<g1::affine_element>(num_points);
    for (size_t i = 0; i < num_points; ++i) {
        points[i] = from_buffer<g1::affine_element>(data, i * sizeof(g1::affine_element));
    }
    return points;
}

g2::affine_element get_bn254_g2_data(const std::filesystem::path& path, bool allow_download)
{
    std::filesystem::create_directories(path);

    auto g2_path = path / "bn254_g2.dat";
    size_t g2_file_size = get_file_size(g2_path);

    if (g2_file_size == sizeof(g2::affine_element)) {
        auto data = read_file(g2_path);
        return from_buffer<g2::affine_element>(data.data());
    }
    if (!allow_download) {
        throw_or_abort("bn254 g2 data not found and download not allowed in this context");
    }
    auto data = download_bn254_g2_data();
    write_file(g2_path, data);
    return from_buffer<g2::affine_element>(data.data());
}
} // namespace bb
