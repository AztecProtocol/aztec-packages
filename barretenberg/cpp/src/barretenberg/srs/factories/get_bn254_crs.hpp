#pragma once
#include <barretenberg/ecc/curves/bn254/g1.hpp>
#include <barretenberg/ecc/curves/bn254/g2.hpp>
#include <filesystem>
#include <fstream>
#include <ios>

namespace bb {
std::vector<g1::affine_element> get_bn254_g1_data(const std::filesystem::path& path,
                                                  size_t num_points,
                                                  bool allow_download = true);
g2::affine_element get_bn254_g2_data(const std::filesystem::path& path, bool allow_download = true);
} // namespace bb
