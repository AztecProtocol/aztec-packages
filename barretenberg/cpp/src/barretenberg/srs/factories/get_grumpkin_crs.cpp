#include "get_grumpkin_crs.hpp"
#include "barretenberg/api/exec_pipe.hpp"
#include "barretenberg/api/file_io.hpp"
#include "barretenberg/common/throw_or_abort.hpp"
#include "barretenberg/common/try_catch_shim.hpp"
#include "barretenberg/ecc/curves/grumpkin/grumpkin.hpp"

namespace {
std::vector<uint8_t> download_grumpkin_g1_data(size_t num_points)
{
    size_t g1_end = num_points * sizeof(bb::curve::Grumpkin::AffineElement) - 1;
    std::string url = "https://crs.aztec.network/grumpkin_g1.dat";

    // IMPORTANT: this currently uses a shell, DO NOT let user-controlled strings here.
    std::string command = "curl -s -H \"Range: bytes=0-" + std::to_string(g1_end) + "\" '" + url + "'";

    auto data = bb::exec_pipe(command);
    if (data.size() < g1_end) {
        THROW std::runtime_error("Failed to download grumpkin g1 data.");
    }

    return data;
}
} // namespace

namespace bb {
std::vector<curve::Grumpkin::AffineElement> get_grumpkin_g1_data(const std::filesystem::path& path,
                                                                 size_t num_points,
                                                                 bool allow_download)
{
    // TODO(AD): per Charlie this should just download and replace the flat file portion atomically so we have no race
    // condition
    std::filesystem::create_directories(path);
    auto g1_path = path / "grumpkin_g1.flat.dat";
    size_t g1_downloaded_points = get_file_size(g1_path) / sizeof(curve::Grumpkin::AffineElement);
    if (g1_downloaded_points >= num_points) {
        vinfo("using cached grumpkin crs with num points ", g1_downloaded_points, " at: ", g1_path);
        auto data = read_file(g1_path, num_points * sizeof(curve::Grumpkin::AffineElement));
        std::vector<curve::Grumpkin::AffineElement> points(num_points);
        for (uint32_t i = 0; i < num_points; ++i) {
            points[i] = from_buffer<curve::Grumpkin::AffineElement>(data, i * sizeof(curve::Grumpkin::AffineElement));
        }
        if (points[0].on_curve()) {
            return points;
        }
    }
    if (!allow_download && g1_downloaded_points == 0) {
        throw_or_abort("grumpkin g1 data not found and download not allowed in this context");
    } else if (!allow_download) {
        throw_or_abort(format("grumpkin g1 data had ",
                              g1_downloaded_points,
                              " points and ",
                              num_points,
                              " were requested but download not allowed in this context"));
    }
    vinfo("downloading grumpkin crs...");
    auto data = download_grumpkin_g1_data(num_points);
    write_file(path / "grumpkin_g1.flat.dat", data);

    std::vector<curve::Grumpkin::AffineElement> points(num_points);
    for (uint32_t i = 0; i < num_points; ++i) {
        points[i] = from_buffer<curve::Grumpkin::AffineElement>(data, i * sizeof(curve::Grumpkin::AffineElement));
    }
    return points;
}
} // namespace bb
