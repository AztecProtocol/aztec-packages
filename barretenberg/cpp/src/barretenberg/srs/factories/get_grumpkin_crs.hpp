#pragma once
#include <barretenberg/ecc/curves/bn254/g1.hpp>
#include <barretenberg/ecc/curves/grumpkin/grumpkin.hpp>
#include <filesystem>
#include <fstream>
#include <ios>

namespace bb {

std::vector<curve::Grumpkin::AffineElement> get_grumpkin_g1_data(const std::filesystem::path& path,
                                                                 size_t num_points,
                                                                 bool allow_download = true);

}
