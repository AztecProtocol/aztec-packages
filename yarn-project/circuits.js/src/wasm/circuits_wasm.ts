import { numToUInt32BE } from '@aztec/foundation/serialize';
import { fileURLToPath } from '@aztec/foundation/url';
import { IWasmModule, WasmModule } from '@aztec/foundation/wasm';

import isNode from 'detect-node';
import { readFile } from 'fs/promises';
import { dirname, join } from 'path';

import { Crs } from '../crs/index.js';

const NAME = '/aztec3-circuits';
const CODE_PATH = isNode
  ? join(dirname(fileURLToPath(import.meta.url)), `../../resources/${NAME}.wasm`)
  : `${NAME}.wasm`;
const MAX_SRS_POINTS = 50000;

/**
 * Get the text of a binary file, either locally or on the web.
 * @param path - Path to the WASM binary.
 * @returns The binary buffer.
 */
async function fetchBinary(path: string) {
  if (isNode) {
    return await readFile(path);
  } else {
    const res = await fetch(path);
    return Buffer.from(await res.arrayBuffer());
  }
}
const BB_JS_NYI_ERROR =
  'NOT YET IMPLEMENTED - needed for proofs, plan is to use barretenberg.js from NPM for proofs. See https://github.com/AztecProtocol/aztec-packages/issues/781';
/**
 * A low-level wrapper for an instance of Aztec3 Circuits WASM.
 */
export class CircuitsWasm implements IWasmModule {
  static instance: Promise<CircuitsWasm>;

  /**
   * Creates a Circuits module from a properly configured WasmModule.
   * Not meant for public use.
   *
   * @param wasm - The module configured in CircuitsWasm.new().
   */
  private constructor(private wasm: WasmModule) {}

  /**
   * Get a singleton instance of the module.
   * @returns The singleton.
   */
  public static get(): Promise<CircuitsWasm> {
    if (!this.instance) this.instance = CircuitsWasm.new();
    return this.instance;
  }

  /**
   * Create and initialize a Circuits module. Not meant for public use.
   *
   * Has 30 pages by default. 30*2**16 \> 1mb stack size plus other overheads.
   * 8192 maximum by default. 512mb.
   * @param initial - Initial memory pages.
   * @param maximum - Max memory pages.
   * @param loggerName - The logger name.
   * @returns The wrapper.
   */
  private static async new(initial = 30, maximum = 8192, loggerName = 'aztec:wasm'): Promise<CircuitsWasm> {
    const wasm = new WasmModule(
      await fetchBinary(CODE_PATH),
      module => ({
        /**
         * Log a string from wasm.
         * @param addr - The string address to log.
         */
        logstr(addr: number) {
          const rawStr = wasm.getMemoryAsString(addr);
          const m = wasm.getMemory();
          const str = `${rawStr} (mem: ${(m.length / (1024 * 1024)).toFixed(2)}MB)`;
          if (str.startsWith('abort: ') || str.startsWith('important: ')) {
            // we explicitly want to route to console.log for every abort message to not miss them:
            // eslint-disable-next-line no-console
            console.log(str);
          }
          wasm.getLogger()(str);
        },
        memory: module.getRawMemory(),
        // eslint-disable-next-line camelcase
        set_data: () => {
          wasm.getLogger()('set_data: NYI');
        },
        // eslint-disable-next-line camelcase
        get_data: () => {
          throw new Error(BB_JS_NYI_ERROR);
        },
      }),
      loggerName,
    );
    await wasm.init(initial, maximum);
    await CircuitsWasm.initializeSrs(wasm);
    return new CircuitsWasm(wasm);
  }

  /**
   * Ensures the global SRS is initialized.
   * Currently used in VK serialization and will be used in proofs.
   * TODO(AD): proof should use external bb.js
   * TODO(AD): revisit when SRS should be initialized
   * @param wasm - The WASM module.
   */
  private static async initializeSrs(wasm: WasmModule) {
    const crs = new Crs(MAX_SRS_POINTS);
    await crs.init();
    const g1Buf = wasm.call('bbmalloc', crs.getG1Data().length);
    wasm.writeMemory(g1Buf, crs.getG1Data());
    const g1SizeBuf = wasm.call('bbmalloc', 4);
    wasm.writeMemory(g1SizeBuf, numToUInt32BE(crs.numPoints));
    const g2Buf = wasm.call('bbmalloc', crs.getG2Data().length);
    wasm.writeMemory(g2Buf, crs.getG2Data());
    wasm.call('srs_init_srs', g1Buf, g1SizeBuf, g2Buf);
    wasm.call('bbfree', g1Buf);
    wasm.call('bbfree', g1SizeBuf);
    wasm.call('bbfree', g2Buf);
  }

  /**
   * Retrieve the exports object of the CircuitsWasm module.
   *
   * @returns An object containing exported functions and properties.
   */
  public exports(): any {
    return this.wasm.exports();
  }

  /**
   * Get a slice of memory between two addresses.
   * @param start - The start address.
   * @param end - The end address.
   * @returns A Uint8Array view of memory.
   */
  public getMemorySlice(start: number, end: number) {
    return this.wasm.getMemorySlice(start, end);
  }

  /**
   * Write data into the heap.
   * @param arr - The data to write.
   * @param offset - The address to write data at.
   */
  public writeMemory(offset: number, arr: Uint8Array) {
    this.wasm.writeMemory(offset, arr);
  }

  /**
   * Get memory as string.
   * @param offset - The address to get null-terminated string data from.
   * @returns JS string.
   */
  public getMemoryAsString(offset: number) {
    return this.wasm.getMemoryAsString(offset);
  }

  /**
   * Calls into the WebAssembly.
   * @param name - The method name.
   * @param args - The arguments to the method.
   * @returns The numeric integer or address result.
   */
  public call(name: string, ...args: any): number {
    return this.wasm.call(name, ...args);
  }
}
