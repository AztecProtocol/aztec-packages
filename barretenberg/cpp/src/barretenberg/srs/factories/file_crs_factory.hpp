#pragma once
#include "../io.hpp"
#include "barretenberg/ecc/curves/bn254/bn254.hpp"
#include "barretenberg/ecc/curves/grumpkin/grumpkin.hpp"
#include "barretenberg/ecc/scalar_multiplication/point_table.hpp"
#include "barretenberg/ecc/scalar_multiplication/scalar_multiplication.hpp"
#include "crs_factory.hpp"
#include <cstddef>
#include <utility>

namespace bb::srs::factories {

/**
 * Create reference strings given a path to a directory of transcript files.
 */
template <typename Curve> class FileCrsFactory : public CrsFactory<Curve> {
  public:
    FileCrsFactory(std::string path, size_t initial_degree = 0);
    FileCrsFactory(FileCrsFactory&& other) = default;

    std::shared_ptr<bb::srs::factories::ProverCrs<Curve>> get_prover_crs(size_t degree) override;

    std::shared_ptr<bb::srs::factories::VerifierCrs<Curve>> get_verifier_crs(size_t degree = 0) override;

  private:
    std::string path_;
    size_t degree_;
    std::shared_ptr<bb::srs::factories::ProverCrs<Curve>> prover_crs_;
    std::shared_ptr<bb::srs::factories::VerifierCrs<Curve>> verifier_crs_;
};

template <typename Curve> class FileProverCrs : public ProverCrs<Curve> {
  public:
    FileProverCrs(const size_t num_points, std::string const& path)
        : num_points(num_points)
    {
        monomials_ = scalar_multiplication::point_table_alloc<typename Curve::AffineElement>(num_points);

        srs::IO<Curve>::read_transcript_g1(monomials_.get(), num_points, path);
        scalar_multiplication::generate_pippenger_point_table<Curve>(monomials_.get(), monomials_.get(), num_points);
    };

    typename Curve::AffineElement* get_monomial_points() { return monomials_.get(); }

    [[nodiscard]] size_t get_monomial_size() const { return num_points; }

  private:
    size_t num_points;
    std::shared_ptr<typename Curve::AffineElement[]> monomials_;
};

template <typename Curve> class FileVerifierCrs : public VerifierCrs<Curve> {
  public:
    FileVerifierCrs(std::string const& path, const size_t num_points);
};

template <> class FileVerifierCrs<curve::BN254> : public VerifierCrs<curve::BN254> {
    using Curve = curve::BN254;

  public:
    FileVerifierCrs(std::string const& path, const size_t num_points = 0);
    virtual ~FileVerifierCrs();
    Curve::G2AffineElement get_g2x() const override { return g2_x; };
    pairing::miller_lines const* get_precomputed_g2_lines() const override { return precomputed_g2_lines; };
    Curve::AffineElement get_first_g1() const override { return first_g1; };

  private:
    Curve::AffineElement first_g1;
    Curve::G2AffineElement g2_x;
    pairing::miller_lines* precomputed_g2_lines;
};

template <> class FileVerifierCrs<curve::Grumpkin> : public VerifierCrs<curve::Grumpkin> {
    using Curve = curve::Grumpkin;

  public:
    FileVerifierCrs(std::string const& path, const size_t num_points);
    virtual ~FileVerifierCrs() = default;
    Curve::AffineElement* get_monomial_points() const override;
    size_t get_monomial_size() const override;
    Curve::AffineElement get_first_g1() const override { return first_g1; };

  private:
    Curve::AffineElement first_g1;
    size_t num_points;
    std::shared_ptr<Curve::AffineElement[]> monomials_;
};

} // namespace bb::srs::factories
