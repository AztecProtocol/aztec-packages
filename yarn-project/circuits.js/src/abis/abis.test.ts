import times from 'lodash.times';

import { AztecAddress, Fr, FunctionData, FunctionLeafPreimage, FunctionSelector, NewContractData } from '../index.js';
import { makeAztecAddress, makeEthAddress, makePoint, makeTxRequest, makeVerificationKey } from '../tests/factories.js';
import { CircuitsWasm } from '../wasm/circuits_wasm.js';
import {
  computeCommitmentNonce,
  computeContractAddress,
  computeContractLeaf,
  computeFunctionLeaf,
  computeFunctionSelector,
  computeFunctionTreeRoot,
  computeUniqueCommitment,
  computeVarArgsHash,
  hashConstructor,
  hashTxRequest,
  hashVK,
  siloCommitment,
  siloNullifier,
} from './abis.js';

describe('abis wasm bindings', () => {
  let wasm: CircuitsWasm;
  beforeAll(async () => {
    wasm = await CircuitsWasm.get();
  });

  it('hashes a tx request', () => {
    const txRequest = makeTxRequest();
    const hash = hashTxRequest(wasm, txRequest);
    expect(hash).toMatchSnapshot();
  });

  it('computes a function selector', () => {
    const funcSig = 'transfer(address,uint256)';
    const res = computeFunctionSelector(wasm, funcSig);
    expect(res).toMatchSnapshot();
  });

  it('hashes VK', () => {
    const vk = makeVerificationKey();
    const res = hashVK(wasm, vk.toBuffer());
    expect(res).toMatchSnapshot();
  });

  it('computes a function leaf', () => {
    const leaf = new FunctionLeafPreimage(new FunctionSelector(7837), false, true, Fr.ZERO, Fr.ZERO);
    const res = computeFunctionLeaf(wasm, leaf);
    expect(res).toMatchSnapshot();
  });

  it('computes function tree root', () => {
    const res = computeFunctionTreeRoot(wasm, [new Fr(0n), new Fr(0n), new Fr(0n), new Fr(0n)]);
    expect(res).toMatchSnapshot();
  });

  it('hashes constructor info', () => {
    const functionData = new FunctionData(FunctionSelector.empty(), false, true, true);
    const argsHash = new Fr(42);
    const vkHash = Buffer.alloc(32);
    const res = hashConstructor(wasm, functionData, argsHash, vkHash);
    expect(res).toMatchSnapshot();
  });

  it('computes a contract address', () => {
    const deployerPubKey = makePoint();
    const contractAddrSalt = new Fr(2n);
    const treeRoot = new Fr(3n);
    const constructorHash = new Fr(4n);
    const res = computeContractAddress(wasm, deployerPubKey, contractAddrSalt, treeRoot, constructorHash);
    expect(res).toMatchSnapshot();
  });

  it('computes commitment nonce', () => {
    const nullifierZero = new Fr(123n);
    const commitmentIndex = 456;
    const res = computeCommitmentNonce(wasm, nullifierZero, commitmentIndex);
    expect(res).toMatchSnapshot();
  });

  it('computes unique commitment', () => {
    const nonce = new Fr(123n);
    const innerCommitment = new Fr(456);
    const res = computeUniqueCommitment(wasm, nonce, innerCommitment);
    expect(res).toMatchSnapshot();
  });

  it('computes siloed commitment', () => {
    const contractAddress = new AztecAddress(new Fr(123n).toBuffer());
    const uniqueCommitment = new Fr(456);
    const res = siloCommitment(wasm, contractAddress, uniqueCommitment);
    expect(res).toMatchSnapshot();
  });

  it('computes siloed nullifier', () => {
    const contractAddress = new AztecAddress(new Fr(123n).toBuffer());
    const innerNullifier = new Fr(456);
    const res = siloNullifier(wasm, contractAddress, innerNullifier);
    expect(res).toMatchSnapshot();
  });

  it('computes contract leaf', () => {
    const cd = new NewContractData(makeAztecAddress(), makeEthAddress(), new Fr(3n));
    const res = computeContractLeaf(wasm, cd);
    expect(res).toMatchSnapshot();
  });

  it('hashes empty function args', async () => {
    const res = await computeVarArgsHash(wasm, []);
    expect(res).toMatchSnapshot();
  });

  it('hashes function args', async () => {
    const args = times(8, i => new Fr(i));
    const res = await computeVarArgsHash(wasm, args);
    expect(res).toMatchSnapshot();
  });

  it('hashes many function args', async () => {
    const args = times(200, i => new Fr(i));
    const res = await computeVarArgsHash(wasm, args);
    expect(res).toMatchSnapshot();
  });
});
