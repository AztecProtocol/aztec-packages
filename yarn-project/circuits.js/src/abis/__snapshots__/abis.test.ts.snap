// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`abis wasm bindings computes a contract address 1`] = `
AztecAddress {
  "buffer": {
    "data": [
      17,
      160,
      201,
      96,
      127,
      185,
      91,
      146,
      4,
      147,
      169,
      211,
      33,
      226,
      156,
      174,
      240,
      174,
      157,
      149,
      81,
      249,
      149,
      226,
      160,
      18,
      239,
      108,
      218,
      140,
      217,
      137,
    ],
    "type": "Buffer",
  },
}
`;

exports[`abis wasm bindings computes a function leaf 1`] = `
Fr {
  "value": 16977277925454205296890617263907769463254781689473762721237698889614686866498n,
}
`;

exports[`abis wasm bindings computes a function selector 1`] = `
{
  "data": [
    169,
    5,
    156,
    187,
  ],
  "type": "Buffer",
}
`;

exports[`abis wasm bindings computes commitment nonce 1`] = `
Fr {
  "value": 13496027098354931963552650893736311555066207973436318315473061104559483240114n,
}
`;

exports[`abis wasm bindings computes contract leaf 1`] = `
Fr {
  "value": 14753252017424563237038015065402454919037294146339173328174383872613705551896n,
}
`;

exports[`abis wasm bindings computes function tree root 1`] = `
Fr {
  "value": 10347099364184817470749359182712341488544005132594679915946302065868679911885n,
}
`;

exports[`abis wasm bindings computes siloed commitment 1`] = `
Fr {
  "value": 16534789767920606571541683139317465939081284769910575930923078591683356873433n,
}
`;

exports[`abis wasm bindings computes siloed nullifier 1`] = `
Fr {
  "value": 4957445512951137590281746823314455197982530036508465893793736233926551006579n,
}
`;

exports[`abis wasm bindings computes unique commitment 1`] = `
Fr {
  "value": 4300635537850003871451650912699252898033249035618622792407744400807361159458n,
}
`;

exports[`abis wasm bindings hashes VK 1`] = `
{
  "data": [
    18,
    27,
    182,
    124,
    37,
    144,
    241,
    62,
    203,
    104,
    141,
    178,
    22,
    93,
    79,
    221,
    21,
    63,
    245,
    104,
    71,
    249,
    55,
    250,
    39,
    168,
    54,
    232,
    57,
    46,
    130,
    254,
  ],
  "type": "Buffer",
}
`;

exports[`abis wasm bindings hashes a tx request 1`] = `
{
  "data": [
    32,
    25,
    83,
    68,
    20,
    2,
    71,
    30,
    241,
    52,
    99,
    157,
    202,
    119,
    186,
    154,
    22,
    122,
    199,
    176,
    121,
    215,
    85,
    13,
    228,
    82,
    129,
    89,
    153,
    131,
    233,
    177,
  ],
  "type": "Buffer",
}
`;

exports[`abis wasm bindings hashes constructor info 1`] = `
Fr {
  "value": 11423915673915198290300367851881125996395433281184245872633888457016608552774n,
}
`;

exports[`abis wasm bindings hashes empty function args 1`] = `
Fr {
  "value": 0n,
}
`;

exports[`abis wasm bindings hashes function args 1`] = `
Fr {
  "value": 18906676604864959005173276375312600640969342814693694325262988108261136820631n,
}
`;

exports[`abis wasm bindings hashes many function args 1`] = `
Fr {
  "value": 16458590480032755384470108712609389694120470087885514794475677859856835664159n,
}
`;
